### DISCUSSION TITLE
[jest] jest.mocked uses ReturnType, which breaks on overloaded functions

### DISCUSSION THREAD

**Discussion by brandon-leapyear** [DefinitelyTyped/DefinitelyTyped#63994] [Issues with a @types package]
`ReturnType` will only use the last defined overload (https://stackoverflow.com/questions/52760509/typescript-returntype-of-overloaded-function). Perhaps `jest.mocked` should use one of the workarounds in the SO post to work for overloaded functions?

We got this with `jest.mocked(fs.readdirSync)`. Worked around by explicitly specifying the type
```ts
jest.mocked<(dir: string) => string[]>(fs.readdirSync)
```

Not sure if there should be a separate fix in the [node] types for making the `readdir*` overload work with `ReturnType`

[a comment by an automated account is not shown]

**Comment by mrazauskas** [comment]
I think it’s fine that TypeScript needs hints in complex situations. There is no need to make the types more complex.

Perhaps it would make sense to document type argument of `jest.mocked()`? I mean, to add your example as a tip to Jest docs?

**Reply by brandonchinn178** [reply]
Sure; I think an explicit mention of "you need this if the function has an overloaded signature" would help
